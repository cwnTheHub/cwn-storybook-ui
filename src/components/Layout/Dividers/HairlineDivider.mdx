{/* HairlineDivider.mdx */}

import {  Canvas, Meta, Title, Subtitle, Description, Primary, Controls, Stories } from '@storybook/blocks';

import * as HairlineDividerStories from './HairlineDivider.stories';

<Meta of={HairlineDividerStories} />
<Title />
	
`` import { HairlineDivider } from '@nds_core/nds' ``

Separate content within modules.

### Usage
HairlineDivider is used to help separate content. Use Box to add space between the HairlineDivider and content.
<Canvas of={HairlineDividerStories.Horizontal} />

<Canvas of={HairlineDividerStories.Vertical} />

<Canvas of={HairlineDividerStories.Example} />