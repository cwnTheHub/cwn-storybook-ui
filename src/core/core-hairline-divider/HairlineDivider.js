import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { borders, spacing } from "../../shared-styles";
import { colorGreyGainsboro } from "../core-colours/colours";
import { safeRest } from "../../util-helpers";

const baseStyle = { transform: "rotate(-0.00001deg)", flexShrink: 0 };

const horizontalStyle = {
  ...baseStyle,
  width: "100%",
  height: "1px",
};

const verticalStyle = {
  ...baseStyle,
  display: "inline-block",
  width: "1px",
};

const StyledHairlineDivider = styled.hr(
  spacing.noSpacing,
  borders.none,
  (props) => {
    if (props.vertical && props.gradient) {
      return {
        ...verticalStyle,
        "background-image": `
        linear-gradient(0deg, rgba(216, 216, 216, 0) 0%,
        ${colorGreyGainsboro} 12%,
        ${colorGreyGainsboro} 88%,
        rgba(216, 216, 216, 0) 100%)
      `,
      };
    }
    if (props.vertical && !props.gradient) {
      return {
        ...verticalStyle,
        "background-color": colorGreyGainsboro,
      };
    }
    if (!props.vertical && props.gradient) {
      return {
        ...horizontalStyle,
        "background-image": `
        linear-gradient(90deg, rgba(216, 216, 216, 0) 0%,
        ${colorGreyGainsboro} 7%,
        ${colorGreyGainsboro} 93%,
        rgba(216, 216, 216, 0) 100%)
      `,
      };
    }
    return {
      ...horizontalStyle,
      "background-color": colorGreyGainsboro,
    };
  }
);

const HairlineDivider = ({ vertical, gradient, ...rest }) => (
  <StyledHairlineDivider
    {...safeRest(rest)}
    vertical={vertical}
    gradient={gradient}
  />
);

HairlineDivider.propTypes = {
  vertical: PropTypes.bool,
  gradient: PropTypes.bool,
};

HairlineDivider.defaultProps = {
  vertical: false,
  gradient: false,
};

export default HairlineDivider;
