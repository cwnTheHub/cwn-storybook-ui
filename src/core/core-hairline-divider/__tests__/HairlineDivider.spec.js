import React from "react";
import { shallow, render, mount } from "enzyme";

import HairlineDivider from "../HairlineDivider";

describe("HairlineDivider", () => {
  const doShallow = (props = {}) => shallow(<HairlineDivider {...props} />);

  it("renders horizontally", () => {
    const hairlineDivider = render(<HairlineDivider />);

    expect(hairlineDivider).toMatchSnapshot();
  });

  it("renders vertically", () => {
    const hairlineDivider = render(<HairlineDivider vertical />);

    expect(hairlineDivider).toMatchSnapshot();
  });

  it("is an HTMl hr element", () => {
    const hairlineDivider = doShallow();

    expect(hairlineDivider.render()[0].name).toEqual("hr");
  });

  it("can be horizontal or vertical", () => {
    let hairlineDivider = doShallow();
    expect(hairlineDivider.prop("vertical")).toEqual(false);

    hairlineDivider = doShallow({ vertical: true });
    expect(hairlineDivider.prop("vertical")).toEqual(true);
  });

  it("can have a gradient", () => {
    let hairlineDivider = doShallow();
    expect(hairlineDivider.prop("gradient")).toEqual(false);

    hairlineDivider = doShallow({ gradient: true });
    expect(hairlineDivider.prop("gradient")).toEqual(true);
  });

  it("renders when zoomed out on Chrome", () => {
    const hairlineDivider = mount(<HairlineDivider />);

    expect(hairlineDivider).toHaveStyleRule("transform", "rotate(-0.00001deg)");
  });

  it("passes additional attributes to the element", () => {
    const hairlineDivider = doShallow({
      id: "the-id",
      "data-some-attr": "some value",
    });

    expect(hairlineDivider).toHaveProp("id", "the-id");
    expect(hairlineDivider).toHaveProp("data-some-attr", "some value");
  });

  it("does not allow custom CSS", () => {
    const hairlineDivider = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(hairlineDivider).not.toHaveProp("className", "my-custom-class");
    expect(hairlineDivider).not.toHaveProp("style");
  });
});
