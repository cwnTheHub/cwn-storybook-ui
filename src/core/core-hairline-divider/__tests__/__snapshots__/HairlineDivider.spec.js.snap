// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HairlineDivider renders horizontally 1`] = `
<hr
  class="sc-aXZVg jFqjSE"
/>
`;

exports[`HairlineDivider renders vertically 1`] = `
<hr
  class="sc-aXZVg kxIxLo"
/>
`;
